tests/components/mqtt/test_init.py

```diff
@@ -64,619 +58,642 @@ def mock_mqtt():
-        mock_client().subscribe.return_value = (0, 0)
-        mock_client().unsubscribe.return_value = (0, 0)
-        mock_client().publish.return_value = (0, 0)
-        result = await async_setup_component(hass, mqtt.DOMAIN, {mqtt.DOMAIN: config})
-        assert result
-        await hass.async_block_till_done()
-        return mock_client()
 
+@pytest.fixture
+def record_calls(calls):
+    """Fixture to record calls."""
 
-mock_mqtt_client = threadsafe_coroutine_factory(async_mock_mqtt_client)
+    @callback
+    def record_calls(*args):
+        """Record calls."""
+        calls.append(args)
 
+    return record_calls
 
-# pylint: disable=invalid-name
-class TestMQTTComponent(unittest.TestCase):
-    """Test the MQTT component."""
 
-    def setUp(self):  # pylint: disable=invalid-name
-        """Set up things to be run when tests are started."""
-        self.hass = get_test_home_assistant()
-        mock_mqtt_component(self.hass)
-        self.calls = []
-        self.addCleanup(self.tear_down_cleanup)
+async def test_mqtt_connects_on_home_assistant_mqtt_setup(
+    hass, mqtt_client_mock, mqtt_mock
+):
+    """Test if client is connected after mqtt init on bootstrap."""
+    assert mqtt_client_mock.connect.call_count == 1
 
-    def tear_down_cleanup(self):
-        """Stop everything that was started."""
-        self.hass.stop()
 
-    @callback
-    def record_calls(self, *args):
-        """Record calls."""
-        self.calls.append(args)
+async def test_mqtt_disconnects_on_home_assistant_stop(hass, mqtt_mock):
+    """Test if client stops on HA stop."""
+    hass.bus.fire(EVENT_HOMEASSISTANT_STOP)
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    assert mqtt_mock.async_disconnect.called
 
-    def aiohttp_client_stops_on_home_assistant_start(self):
-        """Test if client stops on HA stop."""
-        self.hass.bus.fire(EVENT_HOMEASSISTANT_STOP)
-        self.hass.block_till_done()
-        assert self.hass.data["mqtt"].async_disconnect.called
 
-    def test_publish_calls_service(self):
-        """Test the publishing of call to services."""
-        self.hass.bus.listen_once(EVENT_CALL_SERVICE, self.record_calls)
+async def test_publish_calls_service(hass, mqtt_mock, calls, record_calls):
+    """Test the publishing of call to services."""
+    hass.bus.async_listen_once(EVENT_CALL_SERVICE, record_calls)
 
-        mqtt.publish(self.hass, "test-topic", "test-payload")
+    mqtt.async_publish(hass, "test-topic", "test-payload")
 
-        self.hass.block_till_done()
+    await hass.async_block_till_done()
 
-        assert len(self.calls) == 1
-        assert self.calls[0][0].data["service_data"][mqtt.ATTR_TOPIC] == "test-topic"
-        assert (
-            self.calls[0][0].data["service_data"][mqtt.ATTR_PAYLOAD] == "test-payload"
-        )
+    assert len(calls) == 1
+    assert calls[0][0].data["service_data"][mqtt.ATTR_TOPIC] == "test-topic"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5737505)